### PR TITLE
[Reassociate] Preserve NUW flags after expr tree rewriting

### PR DESCRIPTION
**Pull request by dtcxzyw**
Alive2: https://alive2.llvm.org/ce/z/38KiC_

### CODE REVIEW

File: llvm/lib/Transforms/Scalar/Reassociate.cpp [hunk 1]
```diff
@@ -1225,7 +1234,7 @@ Value *ReassociatePass::RemoveFactorFromExpression(Value *V, Value *Factor) {
     RedoInsts.insert(BO);
     V = Factors[0].Op;
   } else {
-    RewriteExprTree(BO, Factors);
+    RewriteExprTree(BO, Factors, HasNUW);
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/Reassociate.cpp:1237]: Could there be a problem if NeedsNegate is set?

**Review by nikic — COMMENTED** [review]
This looks plausible to me, but I'm not familiar with this code.

File: llvm/lib/Transforms/Scalar/Reassociate.cpp (hunk 1 above)
Comment (dtcxzyw) [reply llvm/lib/Transforms/Scalar/Reassociate.cpp:1237]: I don't think it is an issue.
The transform `-C *nuw X1 *nuw X2 *nuw X3` --> `C * -(X1 *nuw X2 *nuw X3)` is correct.

**Comment by dtcxzyw** [comment]
Sorry, it doesn't hold for `mul` if one of the operands is zero.
Alive2: https://alive2.llvm.org/ce/z/686-w6

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**